nosyscall.8
.\"
.\" SPDX-License-Identifier: BSD-2-Clause
.\"
.\" Copyright (c) 2021-2026 Ruslan Zalata <rz@fabmicro.ru>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd October 4, 2026
.Dt NOSYSCALL 8
.Os
.Sh NAME
.Nm nosyscall
.Nd is a simple tool to skip (fake), reject or block selected syscalls for an
application.
.Sh SYNOPSIS
.Nm
.Op Fl srbtv 
.Ar syscall_number prog ...
.Sh DESCRIPTION
The
.Nm
utility uses 
.Ar ptrace(2)
to run a program
.Ar prog
under supervision skipping (faking), rejecting or blocking execution of given
.Ar syscall_number .
By faking it means a fake success (0) result code will be returned to program
without actually performing requested syscall. Similarly, when rejecting, a
fake (-1) result code is returned to program. When blocking is set and program
calls a matching
.Ar syscall_number ,
its execution is stopped and utility quits with result code (-2) allowing to
setup syscall traps for program under supervision.

.Nm
utility can be used to trace both FreeBSD and Linux programs running under
Linuxulator, hence it's useful for reverse-engineering or for setting work-arounds
over blocks in Linux proprietary binaries. Note, that FreeBSD and Linux have
different system call numbers. FreeBSD syscall numbers can be found in
.Ar /usr/include/sys/syscall.h .

.Pp
The following options are available:
.Bl -tag -width indent
.It Fl r
Reject
.Ar syscall_number ,
return from syscall with -1 error code.
.It Fl s
Skip
.Ar syscall_number ,
return from syscall with 0 error code faking suscess.
.It Fl b
Block
.Ar syscall_number ,
stop execution with status code -2.
.It Fl t
Trace
.Ar syscall_number ,
print details when syscall matches.
.It Fl v
Be verbose, print details on each syscall.
.El 
.Sh NOTES
The
.Ar syscall_number
can be either a positive interger number, or a tuple like
.Ar number:arg=val
where
.Ar number
is matching syscall number,
.Ar arg
defines syscall argument number starting from zero, and
.Ar val
defines integer value of the
.Ar arg .
In this case trap triggers when program calls syscall
.Ar number
with argument
.Ar arg
having value
.Ar val .
.Pp
It is possible to provide up to 256 flags intermixing syscall skipping,
rejecting, blocking and tracing.
.Pp
This utility currently supports only x86 platforms.
.El
.Sh EXIT STATUS
.Bl -tag -width indent
.It -1 on error parsing arguments.
.It -2 when syscall block triggers. 
.El 
.Sh EXAMPLES
To reject open() syscall for /bin/ls:
.Bd -literal -offset indent
\*[Gt] nosyscall -v -r 5 /bin/ls 
.Ed
.Pp
To run linux binary /some/linux/app faking fcntl() syscall when cmd is F_SETLK:
.Bd -literal -offset indent
\*[Gt] nosyscall -v -s 72:1=6 /some/linux/app 
.Ed
.Pp
Basically, the above fakes setting lock on a file.
.Ed
.Pp
.Sh SEE ALSO
.Xr ptrace 2 ,
.Xr syscall 2 ,
.Xr truss 1
.Sh AUTHORS
.An Ruslan Zalata Aq Mt rz@fabmicro.ru


nosyscall.c
/*
 * nosyscall is a simple tool to skip or block selected syscalls which is sometimes
 * needed when running proprietary Linux apps on FreeBSD with the linuxulator,
 * e.g. faking flock() or open() syscalls to work-around some blocks.
 *
 * See also truss(1)
 *
 * Note: this is for X86_AMD64 architecture only. You have to modify register names
 * to get it working on some other architectures.
 *
 * Copyright(C) 2021-2026, Ruslan Zalata <rz@fabmicro.ru>
 *
 * SPDX-License-Identifier: BSD-2-Clause
 *
 */


#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <unistd.h>
#include <sys/ptrace.h>
#include <sys/wait.h>
#include <machine/reg.h>
#include <signal.h>
#include <sys/syscall.h>
#include <sys/sysctl.h>

#define	FREEBSD_SYS_nanosleep	240
#define LINUX_SYS_nanosleep	35

#define	MODE_NONE		0
#define	MODE_SKIP		1
#define	MODE_REJECT		2
#define	MODE_BLOCK		3
#define	MODE_TRACE		4

#define	BLOCK_LIST_SIZE		256

char *modes[] = { "none", "skip", "reject", "block", "trace" };

struct block_syscall {
	int syscall_number;
	int sysarg_number;
	long sysarg_value;
	int mode;
};


struct block_syscall *block_syscall_list = NULL;

int sys_nanosleep = 0;	/* nanosleep() syscall number, depends on ABI */


int
addsyscall(char *str, int mode)
{
	int idx;

	for (idx = 0; idx < BLOCK_LIST_SIZE; idx++)
		if (block_syscall_list[idx].syscall_number == 0)
			break;

	if (idx == BLOCK_LIST_SIZE) {
		fprintf(stderr, "No more room in block list!\n");
		exit(-EXIT_FAILURE);
	}

	block_syscall_list[idx].mode = mode;

	char *str_arg = index(str, ':');

	if (str_arg == NULL) {
		// Just a syscall, no argument provided
		block_syscall_list[idx].syscall_number = strtol(str, NULL, 0);
		block_syscall_list[idx].sysarg_number = -1; /* do not match arguments */
		return (idx);
	}

	*str_arg++ = 0;

	char *str_val = index(str_arg, '=');

	if (str_val == NULL) {
		/* No value provided, use zero */
		block_syscall_list[idx].syscall_number = strtol(str, NULL, 0);
		block_syscall_list[idx].sysarg_number = strtol(str_arg, NULL, 0) % 4; /* only four args supported */
		block_syscall_list[idx].sysarg_value = 0;
		return (idx);
	}

	*str_val++ = 0;

	/* All three provided: syscall number, argument number and argumen value */

	block_syscall_list[idx].syscall_number = strtol(str, NULL, 0);
	block_syscall_list[idx].sysarg_number = strtol(str_arg, NULL, 0);
	block_syscall_list[idx].sysarg_value = strtol(str_val, NULL, 0);

	return (idx);
}

int
getdata(int pid, long addr, long *data)
{
	int status;

	if (data == NULL)
		return (-1);

	if (pid <= 0)
		return (-1);

	errno = 0;

	*data = ptrace(PT_READ_D, pid, (caddr_t) addr, 0);

	return (-errno);
}


int
getregs(int pid, struct reg *registers)
{
	int status;

	if (registers == NULL)
		return (-1);

	if (pid <= 0)
		return (-1);

	status = ptrace(PT_GETREGS, pid, (caddr_t) registers, 0);

	if (status != 0)
		return (-1);

	return (registers->r_rax);
}


int
setregs(int pid, struct reg *registers)
{
	int status;

	if (registers == NULL)
		return (-1);

	if (pid <= 0)
		return (-1);

	return (ptrace(PT_SETREGS, pid, (caddr_t) registers, 0));
}


void
run(char **argv)
{
	int status;

	status = ptrace(PT_TRACE_ME, 0, 0, 0);

	if (status != 0)
		exit(-EXIT_FAILURE);

	execv(argv[0], argv);
}


void
usage(const char *prog)
{
	fprintf(stderr, "Usage:\n"
	  "\t%s [-srbtv] syscall_num:arg_num=arg_val /path/to/program_to_be_traced\n"
	  "\t%s [-srbtv] syscall_number /path/to/program_to_be_traced\n\n"
	  "\t-s - to skip given syscall, will return 0 to tracee\n"
	  "\t-r - to reject given syscall, will return -1 to tracee\n"
	  "\t-b - to block given syscall, will stop execution and quit\n"
	  "\t-t - don't do anything, just trace (print info when syscall is met).\n\n"
	  "\t-v - verbose mode (report each syscall).\n\n"
	  "\tA syscall can be identified by 'syscall_num' number and one of its parameter\n"
	  "\tset by 'arg_num' and value 'arg_val'. There can be no more that 256 blocks set at once.\n\n"
	  "Copyright (C) 2021-2026, Ruslan Zalata <rz@fabmicro.ru>\n\n",
	prog, prog);

	exit(-EXIT_FAILURE);
}


int
main(int argc, char **argv)
{
	int verbose = 0;
	int max_syscall_idx = -1;
	struct reg registers;
	int pid;
	int ch;

	if (argc < 3)
		usage(getprogname());

	block_syscall_list = (struct block_syscall *) malloc(
		sizeof(struct block_syscall) * BLOCK_LIST_SIZE);

	if (block_syscall_list == NULL) {
		fprintf(stderr, "Cannot allocate memory for block list: %s\n",
			strerror(errno));
		exit(-EXIT_FAILURE);
	}

	memset((void *) block_syscall_list, 0,
		sizeof(struct block_syscall) * BLOCK_LIST_SIZE);

	while ((ch = getopt(argc, argv, "vs:r:b:t:?h")) != -1) {
		switch (ch) {
		case 's':
			max_syscall_idx = addsyscall(optarg, MODE_SKIP);
			break;

		case 'r':
			max_syscall_idx = addsyscall(optarg, MODE_REJECT);
			break;

		case 'b':
			max_syscall_idx = addsyscall(optarg, MODE_BLOCK);
			break;

		case 't':
			max_syscall_idx = addsyscall(optarg, MODE_TRACE);
			break;

		case 'v':
			verbose++;
			break;

		default:
			usage(getprogname());
		}
	}
	argc -= optind;
	argv += optind;


	if (verbose) {
		fprintf(stderr, "Syscalls to handle: ");

		for (int i = 0; i < max_syscall_idx + 1; i++)
			if (block_syscall_list[i].syscall_number != 0)
				fprintf(stderr, "syscall=%d,argn=%d,argv=%p,mode=%s ",
					block_syscall_list[i].syscall_number,
					block_syscall_list[i].sysarg_number,
					(void *)block_syscall_list[i].sysarg_value,
					modes[block_syscall_list[i].mode]);

		fprintf(stderr, "\n\n");
	}

	pid = fork();
	if (pid == -1) {
		fprintf(stderr, "Fork failed: %s\n", strerror(errno));
		exit(-EXIT_FAILURE);
	} else if (pid == 0) {
		run(argv);
	} else {
		int count = 0;
		int skipped = 0;
		int rejected = 0;

		if (wait(0) == -1) {
			fprintf(stderr, "Wait failed.\n");
			exit(-EXIT_FAILURE);
		}

		/* Get ABI type */

		int mib[4];
		char abi[32];
		size_t len = sizeof(abi);
		mib[0] = CTL_KERN;
		mib[1] = KERN_PROC;
		mib[2] = KERN_PROC_SV_NAME;
		mib[3] = pid;

		if (sysctl(mib, 4, abi, &len, NULL, 0) < 0) {
			fprintf(stderr, "Failed to get ABI type: %s\n",
				strerror(errno));
			exit(-EXIT_FAILURE);
		}

		if (verbose)
			fprintf(stderr, "PID: %d, ABI: %s\n", pid, abi);

		if (strncmp(abi, "Linux", 5) == 0)
			sys_nanosleep = LINUX_SYS_nanosleep;
		else
			sys_nanosleep = FREEBSD_SYS_nanosleep;

		/* Trace till syscall entry */

		while (ptrace(PT_TO_SCE, pid, (caddr_t) 1, 0) == 0) {
			int call;
			int retval;
			int mode = MODE_NONE;
			long sysargs[4];

			if (wait(0) == -1)
				break;

			call = getregs(pid, &registers);

			if (call == -1)
				break;

			/* Obtain syscall args */

			ptrace(PT_GET_SC_ARGS, pid, (caddr_t) &sysargs,
				sizeof(sysargs));

			/* Check if this syscall has to be skipped */

			for (int i = 0; i < max_syscall_idx + 1; i++) {
				if (block_syscall_list[i].syscall_number == 0)
					continue;

				if (block_syscall_list[i].sysarg_number == -1) {
					if (block_syscall_list[i].syscall_number == call) {
						mode = block_syscall_list[i].mode;
						break;
					}
				} else {
					if (block_syscall_list[i].syscall_number == call &&
					   sysargs[block_syscall_list[i].sysarg_number] ==
					   block_syscall_list[i].sysarg_value) {
						mode = block_syscall_list[i].mode;
						break;
					}
				}
			}

			if (verbose > 1 || (mode == MODE_TRACE)) {

				fprintf(stderr,
					"Count: %05d syscall %3d entry\n\trax: 0x%016lx,"
					" rbx: 0x%016lx, rcx: 0x%016lx, rsp: 0x%016lx\n"
					"\trdi: 0x%016lx, rsi: 0x%016lx, rdx: 0x%016lx,"
					" r10: 0x%016lx\n\tstack: ",
					count, call, registers.r_rax, registers.r_rbx,
					registers.r_rcx, registers.r_rsp, registers.r_rdi,
					registers.r_rsi, registers.r_rdx, registers.r_r10);


				for (int i = 0; i < 4; i++) {
					long data;
					if (getdata(pid, registers.r_rsp + i * 8, &data) < 0)
						break;

					fprintf(stderr, "0x%016lx, ", data);
				}

				fprintf(stderr, "\n\targs: ");

				for (int i = 0; i < sizeof(sysargs) / sizeof(*sysargs); i++)
					fprintf(stderr, "0x%016lx, ", sysargs[i]);

				fprintf(stderr, "\n");
			}


			switch (mode) {
			case MODE_SKIP:
				/*
				 * To skip this call to a syscall we have to call
				 * some other harmless syscall like nanosleep()
				 */
				registers.r_rax = sys_nanosleep;
				registers.r_rdi = 0;
				registers.r_rsi = 0;
				setregs(pid, &registers);

				if (verbose)
					fprintf(stderr, "\tSyscall is to be skipped!\n");

				skipped++;
				break;

			case MODE_REJECT:
				/* Same as to skip */
				registers.r_rax = sys_nanosleep;
				registers.r_rdi = 0;
				registers.r_rsi = 0;
				setregs(pid, &registers);

				if (verbose)
					fprintf(stderr, "\tSyscall is to be rejected!\n");

				rejected++;
				break;

			case MODE_BLOCK:
				if (verbose)
					fprintf(stderr, "\tBlocking this syscall!\n");
				exit(-2);
				break;

			default:
			case MODE_NONE:
			case MODE_TRACE:
				break;

			}

			/* proceed tracing till exit() syscall */
			ptrace(PT_TO_SCX, pid, (caddr_t) 1, 0);

			if (wait(0) == -1)
				break;

			retval = getregs(pid, &registers);

			if (verbose > 1 || (mode && verbose))
				fprintf(stderr, "Count: %05d syscall %3d exit, ret_val = %d\n",
					count, call, retval);

			switch (mode) {
			case MODE_SKIP:
				/* indicate no error to tracee */
				registers.r_rax = 0;
				setregs(pid, &registers);

				if (verbose)
					fprintf(stderr, "\tSyscall skipped!\n\n");
				break;

			case MODE_REJECT:
				/* indicate some error to tracee */
				registers.r_rax = -1;
				setregs(pid, &registers);

				if (verbose)
					fprintf(stderr, "\tSyscall rejected!\n\n");
				break;

			case MODE_BLOCK:
				if (verbose)
					fprintf(stderr,
					  "\tSyscall blocked (we should not get to this!)\n\n");
				exit(2);
				break;

			default:
			case MODE_NONE:
			case MODE_TRACE:
				break;

			}

			count++;
		}

		if (verbose)
			fprintf(stderr, "%d system calls processed, %d skipped and %d rejected.\n",
				count, skipped, rejected);
	}

	return (0);
}
